Open AI API hidden states

kyle-pena-kuzco opened the pull request:
## Motivation

This PR exposes hidden states through the OpenAI API.
Hidden states can now be returned by including `return_hidden_states` in the request body.
This feature is enabled by supplying the flag `--enable-return-hidden-states`.

**The test failure noted in the last submission of this PR is resolved, as explained below.**

This feature supports all varieties of OpenAI generation requests:
* v1/chat/completions and v1/completions
* streaming and non-streaming
* Parallel sampling (n>1)

The previous implementation would re-compile the CUDA graph every time the user switched between `return_hidden_states` True and False at the request level.

This behavior is resolved - changing `return_hidden_states` will no longer trigger a CUDA graph re-capture.


## Modifications

* Adds an engine-level flag to enable returning returning hidden states through the engine, native API, or openai API.
* When this engine-level flag is supplied, if CUDA graph is enabled, the CUDA graph runner is compiled in `CaptureHiddenMode.FULL` on startup (so that a double-capture isn't needed).
* This behavior is completely backwards compatible with using the CUDA graph runner for the TARGET model in speculative decoding: `TARGET_VERIFY` runs with `CaptureHiddenMode.FULL`, and `TARGET_EXTEND` does not use the CUDA graph runner.


## Checklist

- [x] Format your code according to the [Code Formatting with Pre-Commit](https://docs.sglang.ai/references/contribution_guide.html#code-formatting-with-pre-commit).
- [x] Add unit tests as outlined in the [Running Unit Tests](https://docs.sglang.ai/references/contribution_guide.html#running-unit-tests-adding-to-ci).
- [ ] Update documentation / docstrings / example tutorials as needed, according to [Writing Documentation](https://docs.sglang.ai/references/contribution_guide.html#writing-documentation-running-docs-ci).
- [x] Provide throughput / latency benchmark results and accuracy evaluation results as needed, according to [Benchmark and Profiling](https://docs.sglang.ai/references/benchmark_and_profiling.html) and [Accuracy Results](https://docs.sglang.ai/references/accuracy_evaluation.html).
- [ ] For reviewers: If you haven't made any contributions to this PR and are only assisting with merging the main branch, please remove yourself as a co-author when merging the PR.
- [ ] Please feel free to join our Slack channel at https://slack.sglang.ai to discuss your PR.

Commit:
initial implementation - no tests

Commit:
removed turtle import

Commit:
added test coverage

Commit:
implemented /v1/completions hidden states

Commit:
small update

kyle-pena-kuzco committed:
fixed the test_completion and test_chat_completion tests

kyle-pena-kuzco committed:
including adapter.py - mistakenly ommitted from last commit

kyle-pena-kuzco committed:
checkpoint before adding more detailed assert messages

kyle-pena-kuzco committed:
all tests pass relating to returning hidden states pass now

kyle-pena-kuzco committed:
fixed a few remaining issues

kyle-pena-kuzco committed:
removed debugging print statement

kyle-pena-kuzco committed:
hidden states now excluded from api response if None, instead of null

kyle-pena-kuzco committed:
removed mistaken way to write out to file

kyle-pena-kuzco committed:
Merge remote-tracking branch 'upstream/main' into openai-hidden-states

kyle-pena-kuzco committed:
last token only, as was intended

kyle-pena-kuzco committed:
Merge remote-tracking branch 'origin/main' into openai-hidden-states

kyle-pena-kuzco committed:
WIP - needs test suite for spec/non-spec/latency impact, needs debug statement cleanup

kyle-pena-kuzco committed:
found a fix for eagle decoding!  get_model_worker_batch() was overriding the preference of the spec decoder in multiple places.  we need to move this to a completely engine-level parameter to fix.

kyle-pena-kuzco committed:
checkpoint before implementing hack-free capture hidden state override / pr cleanup

kyle-pena-kuzco committed:
commit just prior to cleaning up test suite

kyle-pena-kuzco committed:
removed extra loop in test_openai_server dedicated to testing `return_hidden_states`

kyle-pena-kuzco committed:
significant cleanup continues

kyle-pena-kuzco committed:
cleanup continues

kyle-pena-kuzco committed:
more cleanup

kyle-pena-kuzco committed:
test suite cleanup

kyle-pena-kuzco committed:
test suite cleanup

kyle-pena-kuzco committed:
fixed a few outstanding problems revealed by one last regression test prior to testing on H100

kyle-pena-kuzco committed:
fixed the n > 1 problem in adapter.py for hidden states

kyle-pena-kuzco committed:
re-enabled several test suites for hidden states

kyle-pena-kuzco committed:
small changes to test suite

kyle-pena-kuzco committed:
removed some stray prints

kyle-pena-kuzco committed:
Merge branch 'main' into openai-hidden-states

kyle-pena-kuzco committed:
made emission of fingerprint chunks for /v1/completions not dependent on include_usage flag

kyle-pena-kuzco commented on python/sglang/srt/openai_api/adapter.py:
```diff
@@ -669,6 +673,17 @@ def v1_generate_response(
+        elif (not isinstance(request, list)) and request.return_hidden_states:
+            hidden_states = ret_item["meta_info"].get("hidden_states", None)
+        if hidden_states is not None:
+            hidden_states = hidden_states[1:]  # trim off the prefill
+            hidden_states = (
+                hidden_states[-1] if len(hidden_states) > 0 else []
+            )  # slice out the last token
```
This is intentional.  Interacting with `sgl.Engine` directly allows for the possibility of returning the FULL hidden states for all tokens.  However, when returning through an API, serializing hidden states for every token is excessive and not aligned with any currently known use case (whereas returning the hidden states of the last token is indeed a known use case for supporting TopLOC verification).  The plural `hidden_states` is appropriate because even the last token contains a high dimensionality vector, so it is plural ("states").

kyle-pena-kuzco commented on python/sglang/srt/openai_api/adapter.py:
```diff
@@ -1286,6 +1330,21 @@ def v1_chat_generate_response(
         else:
             choice_logprobs = None
 
+        if isinstance(request, list) and request[idx].return_hidden_states:
+            include_hidden_states = True
+        elif not isinstance(request, list) and request.return_hidden_states:
+            include_hidden_states = True
+        else:
+            include_hidden_states = False
```
> Similar to a previous comment, the logic to determine `include_hidden_states` can be simplified to avoid redundant checks and improve readability.

I think the original code is clearer and less verbose.  If the human reviewers agree with gemini I'm happy to change it.

kyle-pena-kuzco commented:
I ran some basic benchmarks on `3.1-8b-instruct` on an H100, comparing this PR with `main`.  I found no performance degradation.

I would like to run similar benchmarks using EAGLE and EAGLE3 but ran into some difficulties: https://sgl-fru7574.slack.com/archives/C07EFURPNN9/p1748476554089449 


`python3 -m sglang.bench_offline_throughput --model-path meta-llama/Meta-Llama-3.1-8B-Instruct --num-prompts 10 `

(This PR)
====== Offline Throughput Benchmark Result =======
Backend:                                 engine
Successful requests:                     10
Benchmark duration (s):                  4.30
Total input tokens:                      1960
Total generated tokens:                  2774
Last generation throughput (tok/s):      138.46
Request throughput (req/s):              2.32
Input token throughput (tok/s):          455.48
Output token throughput (tok/s):         644.64
Total token throughput (tok/s):          1100.12
==================================================

(Main)
====== Offline Throughput Benchmark Result =======
Backend:                                 engine
Successful requests:                     10
Benchmark duration (s):                  4.30
Total input tokens:                      1960
Total generated tokens:                  2774
Last generation throughput (tok/s):      139.63
Request throughput (req/s):              2.32
Input token throughput (tok/s):          455.33
Output token throughput (tok/s):         644.43
Total token throughput (tok/s):          1099.76
==================================================






`python3 -m sglang.bench_offline_throughput --model-path meta-llama/Meta-Llama-3.1-8B-Instruct --num-prompts 10 --disable-cuda-graph`

(This PR)
====== Offline Throughput Benchmark Result =======
Backend:                                 engine
Successful requests:                     10
Benchmark duration (s):                  5.76
Total input tokens:                      1960
Total generated tokens:                  2774
Last generation throughput (tok/s):      101.42
Request throughput (req/s):              1.74
Input token throughput (tok/s):          340.46
Output token throughput (tok/s):         481.86
Total token throughput (tok/s):          822.32
==================================================

(Main)
====== Offline Throughput Benchmark Result =======
Backend:                                 engine
Successful requests:                     10
Benchmark duration (s):                  5.85
Total input tokens:                      1960
Total generated tokens:                  2774
Last generation throughput (tok/s):      99.46
Request throughput (req/s):              1.71
Input token throughput (tok/s):          334.82
Output token throughput (tok/s):         473.88
Total token throughput (tok/s):          808.70
==================================================




`python3 -m sglang.bench_offline_throughput --model-path meta-llama/Meta-Llama-3.1-8B-Instruct --num-prompts 10 --enable-return-hidden-states`

(This PR)
====== Offline Throughput Benchmark Result =======
Backend:                                 engine
Successful requests:                     10
Benchmark duration (s):                  4.31
Total input tokens:                      1960
Total generated tokens:                  2774
Last generation throughput (tok/s):      138.32
Request throughput (req/s):              2.32
Input token throughput (tok/s):          455.12
Output token throughput (tok/s):         644.13
Total token throughput (tok/s):          1099.26
==================================================


`python3 -m sglang.bench_offline_throughput --model-path meta-llama/Meta-Llama-3.1-8B-Instruct --num-prompts 10 --disable-cuda-graph —enable-return-hidden-states`

(This PR)
====== Offline Throughput Benchmark Result =======
Backend:                                 engine
Successful requests:                     10
Benchmark duration (s):                  5.88
Total input tokens:                      1960
Total generated tokens:                  2774
Last generation throughput (tok/s):      100.39
Request throughput (req/s):              1.70
Input token throughput (tok/s):          333.21
Output token throughput (tok/s):         471.59
Total token throughput (tok/s):          804.79
==================================================

kyle-pena-kuzco committed:
added test_openai_server_hidden_states.py to run_suite.py under per-commit

kyle-pena-kuzco committed:
updated docs.  updated examples for hidden states engine and hidden states server

kyle-pena-kuzco committed:
Merge branch 'main' into openai-hidden-states

kyle-pena-kuzco commented on python/sglang/srt/openai_api/adapter.py:
```diff
@@ -669,6 +673,17 @@ def v1_generate_response(
         else:
             logprobs = None
 
+        hidden_states = None
+        if isinstance(request, list) and request[idx].return_hidden_states:
+            hidden_states = ret_item["meta_info"].get("hidden_states", None)
+        elif (not isinstance(request, list)) and request.return_hidden_states:
+            hidden_states = ret_item["meta_info"].get("hidden_states", None)
```
@zhyncs let me know if you agree with gemini's feedback.  Happy to change it if so.

kyle-pena-kuzco commented:
> I ran some basic benchmarks on `3.1-8b-instruct` on an H100, comparing this PR with `main`. I found no performance degradation.
> 
> I would like to run similar benchmarks using EAGLE and EAGLE3 but ran into some difficulties: https://sgl-fru7574.slack.com/archives/C07EFURPNN9/p1748476554089449

@zhyncs let us know if you have any updates / feedback!

zhyncs commented:
Hi @kyle-pena-kuzco can you merge the latest main

kyle-pena-kuzco committed:
Merge branch 'main' into openai-hidden-states

zhyncs commented:
> > Hi @kyle-pena-kuzco can you merge the latest main
> 
> Done!

@kyle-pena-kuzco Nice. @CatherineSue will help review.

CatherineSue reviewed (commented):
Overall changes in openai_api and tests LGTM. Left some comments and questions for better clarification.
May need another pass on the speculative side changes.

kyle-pena-kuzco commented:
> Overall changes in openai_api and tests LGTM. Left some comments and questions for better clarification. May need another pass on the speculative side changes.

Thank you @CatherineSue - I am addressing your feedback now!

kyle-pena-kuzco committed:
Update python/sglang/srt/openai_api/adapter.py

Agree with this simplification.

Co-authored-by: Chang Su <csu272@usc.edu>

kyle-pena-kuzco commented on python/sglang/srt/openai_api/adapter.py:
```diff
@@ -1454,12 +1515,16 @@ async def generate_stream_resp():
                     text = content["text"]
+                    hidden_states[index] = content["meta_info"].get(
+                        "hidden_states", None
+                    ) or hidden_states.get(index)
```
See earlier comment.

CatherineSue commented on python/sglang/srt/openai_api/adapter.py:
```diff
@@ -780,6 +797,9 @@ async def generate_stream_resp():
                     cached_tokens[index] = content["meta_info"].get("cached_tokens", 0)
+                    hidden_states[index] = content["meta_info"].get(
+                        "hidden_states", None
+                    ) or hidden_states.get(index)
```
ahh basically, if the index has been visited before, `hidden_states[index]` may not be None, so we either update `hidden_states[index]` to `content["meta_info"].get("hidden_states")`, or keep it as the same. 
Am I understanding it correctly? 
I was confused because if index is always different, hidden_state.get(index) will always be None as it has never been visited before.

kyle-pena-kuzco commented on test/srt/test_openai_server_hidden_states.py:
```diff
@@ -0,0 +1,356 @@
+
+
+class TestOpenAIServerWithHiddenStatesEnabled(
+    unittest.TestCase, BaseTestOpenAIServerWithHiddenStates
```
Yes, thank for the catch and the great review.

Commit:
addressed requested fixes with review by sglang team

Commit:
Merge branch 'openai-hidden-states' of github.com:context-labs/sglang into openai-hidden-states

CatherineSue reviewed (approved):
@kyle-pena-kuzco Thanks for resolving all the comments. LGTM. If all CI passes I think we are good to go

kyle-pena-kuzco committed:
Merge branch 'main' into openai-hidden-states

kyle-pena-kuzco commented:
> > @kyle-pena-kuzco Thanks for resolving all the comments. LGTM. If all CI passes I think we are good to go
> 
> @CatherineSue - it looks like there are merge conflicts now. i'll fix these today.

hi @CatherineSue - let me know what you think of the run results.  It's difficult for me to tell if these are genuine regressions because I see other commits / PRs with similar failures.

CatherineSue commented:
cc @zhyncs The CI looks good. Should we merge?

kyle-pena-kuzco committed:
Merge branch 'main' into openai-hidden-states

kyle-pena-kuzco committed:
Merge branch 'main' into openai-hidden-states

kyle-pena-kuzco committed:
Merge branch 'main' into openai-hidden-states

zzh-www commented:
@kyle-pena-kuzco Good job!!! But why it would not return the hidden_state of last token? "return_hidden_states" for me means return all hidden_states.

justadogistaken commented:
@kyle-pena-kuzco Hello, It seems this can get last layer hidden_states only. How can we get the entire hidden states?
